doc/reference/trd104-syscalls.md

```diff
@@ -649,61 +685,168 @@ information can be stored in the kernel and used in management or policy decisio
+the value in `error` is undefined.
 
-6.1.1 Return Type Values
+5.1.3 Command
 ---------------------------------
 
+The subscribe system call has this function prototype:
+
+```c
+typedef struct {
+  syscall_rtype_t type;
+  uint32_t data[3];
+} syscall_return_t;
+
+syscall_return_t command(uint32_t driver, uint32_t command, int data, int arg2);
+```
 
+Because a command can return any failure or success variant, it returns a direct
+mapping of the return registers. `rtype` contains the value of `r0`, while
+`data[0]` contains what was passed in `r1`, `data[1]` contains was passed in `r2`,
+and `data[2]` contains what was passed in `r3`.
 
-6.1.2 Yield Return Type
+5.1.4 Read-Write Allow
 ---------------------------------
 
-6.1.3 Subscribe Return Type
+The read-write allow system call has this function prototype:
+
+```c
+typedef struct {
+  bool success;
+  void* ptr;
+  size_t size;
+  tock_error_t error;
+} allow_rw_return_t;
+
+allow_rw_return_t allow_readwrite(uint32_t driver, uint32_t allow, void* ptr, size_t size);
+```
+
+The `success` field indicates whether the call succeeded.
+If it failed, the error code is stored in `error`. If it succeeded,
+the value in `error` is undefined. `ptr` and `size` contain the pointer
+and size of the passed buffer.
+
+
+5.1.5 Read-Only Allow
 ---------------------------------
 
-6.1.4 Command  Return Type
+The read-only allow system call has this function prototype:
+
+```c
+typedef struct {
+  bool success;
+  const void* ptr;
+  size_t size;
+  tock_error_t error;
+} allow_ro_return_t;
+
+allow_ro_return_t allow_readonly(uint32_t driver, uint32_t allow, const void* ptr, size_t size);
+```
+
+The `success` field indicates whether the call succeeded.
+If it failed, the error code is stored in `error`. If it succeeded,
+the value in `error` is undefined. `ptr` and `size` contain the pointer
+and size of the passed buffer.
+
+5.1.6 Memop
 ---------------------------------
 
-6.1.5 Allow Return Type
+Because the Memop system calls are defined by the kernel and not extensible, they are
+directly defined by libtock-c as library functions:
+
+```c
+void* tock_app_memory_begins_at(void);
+void* tock_app_memory_ends_at(void);
+void* tock_app_flash_begins_at(void);
+void* tock_app_flash_ends_at(void);
+void* tock_app_grant_begins_at(void);
+int tock_app_number_writeable_flash_regions(void);
+void* tock_app_writeable_flash_region_begins_at(int region_index);
+void* tock_app_writeable_flash_region_ends_at(int region_index);
+```
+
+They wrap around an underlying function which uses inline assembly:
+
+```c
+void* memop(uint32_t op_type, int arg1);
+```
+
+5.1.7 Exit
 ---------------------------------
 
-6.1.6 Memop Return Type
+The Exit system calls have these function prototypes:
+
+```c
+void tock_exit(uint32_t completion_code);
+void tock_restart(uint32_t completion_code);
+```
+
+Since these two variants of Exit never return, they have
+no return value.
+
+
+5.2 libtock-rs
```

Comment:
`libtock-rs` will have a more complex syscalls API than `libtock-c`, as there will be multiple variations of Subscribe, Read-Only Allow, and Read-Write Allow with different lifetime semantics. I also expect it to keep changing for a while.

Should we get rid of the `libtock-rs` heading?

Reply:
Sure -- the goal here was to show examples. Having a follow-up TRD that specifies libtock-rs would be fine.